File: SiliconLabsApp/BluetoothControllers/SILDiscoveredPeripheral.m
//
//  SILDiscoveredPeripheral.m
//  SiliconLabsApp
//
//  Created by Colden Prime on 1/14/15.
//  Copyright (c) 2015 SiliconLabs. All rights reserved.
//

#import "SILDiscoveredPeripheral.h"
#import "SILRSSIMeasurementTable.h"
#import "SILBluetoothBrowser+Constants.h"
#import "SILConstants.h"

@interface SILDiscoveredPeripheral ()

@property (strong, nonatomic) NSMutableArray *RSSIMeasurements;
@property (nonatomic) double lastTimestamp;
@property (nonatomic) long long packetReceivedCount;

@end

@implementation SILDiscoveredPeripheral

NSString* const RSSIAppendingString = @" RSSI";
NSString* const ConnectableDevice = @"Connectable";
NSString* const NonConnectableDevice = @"Non-connectable";

+ (NSString *)connectableDevice { return ConnectableDevice; }
+ (NSString *)nonConnectableDevice { return NonConnectableDevice; }

- (instancetype)initWithPeripheral:(CBPeripheral *)peripheral
                 advertisementData:(NSDictionary *)advertisementData
                              RSSI:(NSNumber *)RSSI
            andDiscoveringTimestamp:(double)timestamp {
    self = [super init];
    if (self) {
        self.peripheral = peripheral;
        self.uuid = self.peripheral.identifier;
        self.identityKey = [SILDiscoveredPeripheralIdentifierProvider provideKeyForCBPeripheral:self.peripheral];
        self.RSSIMeasurementTable = [[SILRSSIMeasurementTable alloc] init];
        self.advertisingInterval = 0;
        self.packetReceivedCount = 0;
        self.isConnectable = NO;
        self.lastTimestamp = timestamp;
        [self updateWithAdvertisementData:advertisementData RSSI:RSSI andDiscoveringTimestamp:timestamp];
        self.advertisedLocalName = DefaultDeviceName;
    }
    return self;
}

- (void)updateWithAdvertisementData:(NSDictionary *)advertisementData
                               RSSI:(NSNumber *)RSSI
        andDiscoveringTimestamp:(double)timestamp {
    self.advertisedLocalName = advertisementData[CBAdvertisementDataLocalNameKey] ?: self.peripheral.name;
    self.advertisedServiceUUIDs = advertisementData[CBAdvertisementDataServiceUUIDsKey];
    self.txPowerLevel = advertisementData[CBAdvertisementDataTxPowerLevelKey];
    if (!self.isConnectable) {
        self.isConnectable = [advertisementData[CBAdvertisementDataIsConnectable] boolValue];
    }
    self.manufacturerData = advertisementData[CBAdvertisementDataManufacturerDataKey];
    self.beacon = [self parseBeaconData:advertisementData];
    if ([self isCorrectAdvertisingPacket:timestamp]) {
        self.packetReceivedCount++;
        [self calculateAdvertisingIntervalWith:timestamp];
        self.lastTimestamp = timestamp;
    }
    [self.RSSIMeasurementTable addRSSIMeasurement:RSSI];
    [self.delegate peripheral:self didUpdateWithAdvertisementData:advertisementData andRSSI:RSSI];
}

- (BOOL)isCorrectAdvertisingPacket:(double)currentTimestamp {
    double currentInterval = currentTimestamp - self.lastTimestamp;
    if (currentInterval < 0.005) {
        return NO;
    }
    return YES;
}

- (void)calculateAdvertisingIntervalWith:(double)currentTimestamp {
    const double Add3MsToInterval = 0.003;
    const double MultiplierForBegginingCalculating = 0.0007;
    const double MultiplierForAverageInterval = 0.0014;
    const int ReliableForCalculatingPacketAmount = 29;
    const int MinimumCount = 10;
    
    double currentInterval = currentTimestamp - self.lastTimestamp;

    if (currentInterval <= 0) {
        return;
    }
    
    if (self.advertisingInterval == 0) {
        self.advertisingInterval = currentInterval;
    } else if ((currentInterval < self.advertisingInterval * MultiplierForBegginingCalculating) && self.packetReceivedCount < MinimumCount) {
        self.advertisingInterval = currentInterval;
    } else if (currentInterval < self.advertisingInterval + Add3MsToInterval) {
        long long limitedCount = MIN(self.packetReceivedCount, MinimumCount);
        self.advertisingInterval = (((self.advertisingInterval * (limitedCount - 1) + currentInterval)) / limitedCount);
    } else if (currentInterval < self.advertisingInterval * MultiplierForAverageInterval) {
        self.advertisingInterval = (((self.advertisingInterval * ReliableForCalculatingPacketAmount + currentInterval)) / (ReliableForCalculatingPacketAmount + 1));
    }
}

- (void)resetLastTimestampValue {
    self.lastTimestamp = 0.0;
}

- (SILBeacon*)parseBeaconData:(NSDictionary*)adverisement {
    if (self.manufacturerData) {
        NSError* error = nil;
        SILBeacon* beacon = [SILBeacon beaconWithAdvertisment:adverisement name:self.advertisedLocalName error:&error];
        if (error == nil) {
            return beacon;
        }
    }
    
    SILBeacon* unknownBeacon = [[SILBeacon alloc] init];
    unknownBeacon.name = @"Unspecified";
    unknownBeacon.type = SILBeaconTypeUnspecified;
    
    return unknownBeacon;
}

- (instancetype)initWithIBeacon:(CLBeacon*)iBeacon andDiscoveringTimestamp:(double)timestamp {
    self = [super init];
    if (self) {
        self.peripheral = nil;
        if (@available(iOS 13.0, *)) {
            self.uuid = iBeacon.UUID;
        } else {
            self.uuid = iBeacon.proximityUUID;
        }
        self.identityKey = [SILDiscoveredPeripheralIdentifierProvider provideKeyForCLBeacon:iBeacon];
        self.beacon = [SILBeacon beaconWithIBeacon:iBeacon];
        self.beacon.name = SILBeaconIBeacon;
        self.beacon.beacon = iBeacon;
        self.RSSIMeasurementTable = [[SILRSSIMeasurementTable alloc] init];
        self.advertisingInterval = 0;
        self.packetReceivedCount = 0;
        self.isConnectable = NO;
        self.lastTimestamp = timestamp;
        [self updateWithIBeacon:iBeacon andDiscoveringTimestamp:timestamp];
        self.advertisedLocalName = DefaultDeviceName;
    }
    return self;
}

- (void)updateWithIBeacon:(CLBeacon*)iBeacon andDiscoveringTimestamp:(double)timestamp {
    self.advertisedServiceUUIDs = nil;
    self.txPowerLevel = nil;
    self.isConnectable = NO;
    self.manufacturerData = nil;
    if ([self isCorrectAdvertisingPacket:timestamp]) {
        self.packetReceivedCount++;
        [self calculateAdvertisingIntervalWith:timestamp];
        self.lastTimestamp = timestamp;
    }
    NSNumber* rssi = [NSNumber numberWithLong:iBeacon.rssi];
    [self.RSSIMeasurementTable addRSSIMeasurement:rssi];
    [self.delegate peripheral:self didUpdateWithAdvertisementData:nil andRSSI:rssi];
}

- (BOOL)isDMPConnectedLightConnect {
    return [self isContainService:SILServiceNumberConnectedLightingConnect];
}

- (BOOL)isDMPConnectedLightProprietary {
    return [self isContainService:SILServiceNumberConnectedLightingProprietary];
}

- (BOOL)isDMPConnectedLightThread {
    return [self isContainService:SILServiceNumberConnectedLightingThread];
}

- (BOOL)isDMPConnectedLightZigbee {
    return [self isContainService:SILServiceNumberConnectedLightingZigbee];
}

- (BOOL)isRangeTest {
    return [self isContainService:SILServiceNumberRangeTest];
}

- (BOOL)isContainService:(NSString *)serviceUUID {
    CBUUID * const service = [CBUUID UUIDWithString:serviceUUID];
    return [self.advertisedServiceUUIDs containsObject:service];
}

- (NSString *)rssiDescription {
    NSString *rssi = [self.RSSIMeasurementTable.lastRSSIMeasurement stringValue];
    NSMutableString* rssiDescription = [[NSMutableString alloc] initWithString:rssi];
    [rssiDescription appendString:RSSIAppendingString];
    return rssiDescription;
}

@end

File: SiliconLabsApp/BluetoothControllers/SILDiscoveredPeripheral.swift
//
//  SILDiscoveredPeripheral.swift
//  BlueGecko
//
//  Created by Grzegorz Janosz on 09/03/2022.
//  Copyright © 2022 SiliconLabs. All rights reserved.
//

import Foundation
import CoreBluetooth

protocol SILDiscoveredPeripheralDelegate: class {
    func peripheral(_ peripheral: SILDiscoveredPeripheral, didUpdateWithAdvertisementData dictionary: [String: Any]?, andRSSI: NSNumber)
}

@objcMembers
class SILDiscoveredPeripheral: NSObject {
    
    fileprivate let RSSIAppendingString = " RSSI"
    fileprivate let EddystoneService = "FEAA"
    
    var identityKey: String
    var uuid: UUID
    var beacon: SILBeacon!
    var peripheral: CBPeripheral?
    var rssiMeasurementTable = SILRSSIMeasurementTable()
    
    var advertisedLocalName: String?
    var advertisedServiceUUIDs: [CBUUID]?
    var txPowerLevel: NSNumber?
    var manufacturerData: NSData?
    var solicitedServiceUUIDs: [CBUUID]?
    var dataServiceData: [CBUUID : NSData]?
    var advertisingInterval = 0.0
    weak var delegate: SILDiscoveredPeripheralDelegate?
    
    var isFavourite = false
    var isConnectable = false
    
    private var lastTimestamp = 0.0
    private var packetReceivedCount: Int64 = 0
    
    var isDMPConnectedLightConnect: Bool {
        return isContainService(SILServiceNumberConnectedLightingConnect)
    }

    var isDMPConnectedLightProprietary: Bool {
        return isContainService(SILServiceNumberConnectedLightingProprietary)
    }
    
    var isDMPConnectedLightThread: Bool {
        return isContainService(SILServiceNumberConnectedLightingThread)
    }
    
    var isDMPConnectedLightZigbee: Bool {
        return isContainService(SILServiceNumberConnectedLightingZigbee)
    }
    
    var isRangeTest: Bool {
        return isContainService(SILServiceNumberRangeTest)
    }
    
    var hasEddystoneService: Bool {
        return isContainService(EddystoneService)
    }
    
    @objc(initWithPeripheral:advertisementData:RSSI:andDiscoveringTimestamp:)
    init(peripheral: CBPeripheral, advertisementData: [String : Any], rssi RSSI: NSNumber, andDiscoveringTimestamp timestamp: Double
    ) {
        identityKey = SILDiscoveredPeripheralIdentifierProvider.provideKeyForCBPeripheral(peripheral)
        uuid = peripheral.identifier
        super.init()
        self.peripheral = peripheral
        commonInit(timestamp: timestamp)
        update(withAdvertisementData: advertisementData, rssi: RSSI, andDiscoveringTimestamp: timestamp)
    }
    
    init(iBeacon: CLBeacon, andDiscoveringTimestamp timestamp: Double) {
        identityKey = SILDiscoveredPeripheralIdentifierProvider.provideKeyForCLBeacon(iBeacon)
        if #available(iOS 13.0, *) {
            uuid = iBeacon.uuid
        } else {
            uuid = iBeacon.proximityUUID
        }
        beacon = SILBeacon(iBeacon: iBeacon)
        beacon.name = SILBeaconIBeacon
        super.init()
        peripheral = nil
        commonInit(timestamp: timestamp)
        update(withIBeacon: iBeacon, andDiscoveringTimestamp: timestamp)
    }
    
    private func commonInit(timestamp: Double) {
        rssiMeasurementTable = SILRSSIMeasurementTable()
        advertisingInterval = 0
        packetReceivedCount = 0
        isConnectable = false
        lastTimestamp = timestamp
        advertisedLocalName = DefaultDeviceName
    }
    
    @objc(updateWithAdvertisementData:RSSI:andDiscoveringTimestamp:)
    func update(withAdvertisementData advertisementData: [String : Any],
        rssi RSSI: NSNumber,
        andDiscoveringTimestamp timestamp: Double
    ) {
        advertisedLocalName = advertisementData[CBAdvertisementDataLocalNameKey] as? String ?? peripheral!.name
        advertisedServiceUUIDs = advertisementData[CBAdvertisementDataServiceUUIDsKey] as? [CBUUID]
        txPowerLevel = advertisementData[CBAdvertisementDataTxPowerLevelKey] as? NSNumber
        if !isConnectable, let isConnectable = advertisementData[CBAdvertisementDataIsConnectable] as? NSNumber {
            self.isConnectable = isConnectable.boolValue
        }
        manufacturerData = advertisementData[CBAdvertisementDataManufacturerDataKey] as? NSData
        solicitedServiceUUIDs = advertisementData[CBAdvertisementDataSolicitedServiceUUIDsKey] as? [CBUUID]
        dataServiceData = advertisementData[CBAdvertisementDataServiceDataKey] as? [CBUUID: NSData]
        beacon = parseBeaconData(advertisementData)
        if isCorrectAdvertisingPacket(timestamp) {
            packetReceivedCount += 1
            calculateAdvertisingInterval(with: timestamp)
            lastTimestamp = timestamp
        }
        rssiMeasurementTable.addRSSIMeasurement(RSSI)
        delegate?.peripheral(self, didUpdateWithAdvertisementData: advertisementData, andRSSI: RSSI)
    }

    private func isCorrectAdvertisingPacket(_ currentTimestamp: Double) -> Bool {
        let currentInterval = currentTimestamp - lastTimestamp
        return currentInterval >= 0.005
    }
    
    private func calculateAdvertisingInterval(with currentTimestamp: Double) {
        let add3MsToInterval: Double = 0.003
        let multiplierForBegginingCalculating: Double = 0.0007
        let multiplierForAverageInterval: Double = 0.0014
        let reliableForCalculatingPacketAmount: Int64 = 29
        let minimumCount: Int64 = 10

        let currentInterval = currentTimestamp - lastTimestamp
        
        if currentInterval <= 0 {
            return
        }

        if advertisingInterval == 0 {
            advertisingInterval = currentInterval
        } else if (currentInterval < advertisingInterval * multiplierForBegginingCalculating) && packetReceivedCount < minimumCount {
            advertisingInterval = currentInterval
        } else if currentInterval < advertisingInterval + add3MsToInterval {
            let limitedCount = min(packetReceivedCount, minimumCount)
            advertisingInterval = (advertisingInterval * Double(limitedCount - 1) + currentInterval) / Double(limitedCount)
        } else if currentInterval < advertisingInterval * multiplierForAverageInterval {
            advertisingInterval = Double(advertisingInterval * Double(reliableForCalculatingPacketAmount) + currentInterval) / Double(reliableForCalculatingPacketAmount + 1)
        }
    }
    
    private func parseBeaconData(_ advertisement: [String : Any]?) -> SILBeacon {
        if let _ = manufacturerData, let beacon = try? SILBeacon(advertisment: advertisement, name: advertisedLocalName) {
            return beacon
        } else if hasEddystoneService, let dataServiceData = dataServiceData, let data = dataServiceData[CBUUID(string: EddystoneService)] as Data? {
            return SILBeacon(eddystone: data)
        }

        let unknownBeacon = SILBeacon()
        unknownBeacon.name = "Unspecified"
        unknownBeacon.type = .unspecified

        return unknownBeacon
    }
    
    func resetLastTimestampValue() {
        lastTimestamp = 0.0
    }

    func update(withIBeacon iBeacon: CLBeacon, andDiscoveringTimestamp timestamp: Double) {
        advertisedServiceUUIDs = nil
        txPowerLevel = nil
        isConnectable = false
        manufacturerData = nil
        if isCorrectAdvertisingPacket(timestamp) {
            packetReceivedCount += 1
            calculateAdvertisingInterval(with: timestamp)
            lastTimestamp = timestamp
        }
        let rssi = NSNumber(value: iBeacon.rssi)
        rssiMeasurementTable.addRSSIMeasurement(rssi)
        delegate?.peripheral(self, didUpdateWithAdvertisementData: nil, andRSSI: rssi)
    }
        
    private func isContainService(_ serviceUUID: String) -> Bool {
        let service = CBUUID(string: serviceUUID)
        guard let advertisedServiceUUIDs = advertisedServiceUUIDs else {
            return false
        }
        return advertisedServiceUUIDs.contains(service)
    }
    
    func rssiDescription() -> String {
        if let lastMeasurement = rssiMeasurementTable.lastRSSIMeasurement()?.stringValue {
            return lastMeasurement + RSSIAppendingString
        } else {
            return "N/A"
        }
    }

    func rssiValue() -> NSNumber? {
        return rssiMeasurementTable.lastRSSIMeasurement()
    }
}
